compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1076,7 +1076,7 @@ trait Applications extends Compatibility {
           case _ =>
         }
         def tryDynamicTypeApply(): Tree = typedFn match {
-          case typedFn: Select if !pt.isInstanceOf[FunProto] => typedDynamicSelect(typedFn, typedArgs, pt)
+          case typedFn: Select if !pt.isInstanceOf[FunProto] => typedDynamicSelect(typedFn, typedArgs.map(untpd.MemoizedTypeTree(_)), pt)
```

Comment:
Let's see whether we find a way to avoid `MemoizedTypeTree`. Why did it not work with `typedArgs` in the first place? What issue gets fixed by wrapping the arguments in `MemoizedTypeTree`?

Reply:
the issue with `typedArgs` was that after
`tree.args: List(Tuple(List(Ident(Int), Ident(Int))))`

gets typed to
`typedArgs: List(AppliedTypeTree(TypeTree[TypeRef(...,class Tuple2)],List(Ident(Int), Ident(Int))))`

it fails when it tries to type check the `AppliedTypeTree` again because `typedAppliedTypeTree` sets `pt` to `AnyTypeConstructorProto`
https://github.com/lampepfl/dotty/blob/04ea25cfe9c77f612964ba306f070ba058ee695c/compiler/src/dotty/tools/dotc/typer/Typer.scala#L1700-L1703

So I wrapped `typedArgs` with `MemoizedTypeTree` so that `typedTypeTree` could know that the args were already type checked and return the cached results instead of an error type. doing a `hasType` check + cast worked as well, but having an explicit memoize  type tree seemed like a safer solution.

Reply:
Thanks for explaining! We actually already have an analogous memoization mechanism in place. It's `untpd.TypedSplice`. Can you try to simply replace `MemoizedTypeTree` with `untpd.TypedSplice`?

Reply:
thanks for the suggestion! i've updated the PR